server.go

```diff
@@ -121,6 +125,124 @@ type updatePeerHeightsMsg struct {
 	originPeer *serverPeer
 }
 
+// naSubmission represents a network address submission from an outbound peer.
+type naSubmission struct {
+	na           *wire.NetAddress
+	netType      addrmgr.NetworkAddress
+	reach        int
+	score        uint32
+	lastAccessed int64
+}
+
+// naSubmissionCache represents a bounded map for network address submisions.
+type naSubmissionCache struct {
+	cache map[string]*naSubmission
+	limit int
+	mtx   sync.Mutex
+}
+
+// add caches the provided address submission.
+func (sc *naSubmissionCache) add(sub *naSubmission) error {
+	sc.mtx.Lock()
+	defer sc.mtx.Unlock()
+
+	if sub == nil {
+		return fmt.Errorf("submission cannot be nil")
+	}
+
+	key := sub.na.IP.String()
+	if key == "" {
+		return fmt.Errorf("submission key cannot be an empty string")
+	}
+
+	// Remove the oldest submission if cache limit has been reached.
+	if len(sc.cache) == sc.limit {
+		var oldestSub *naSubmission
+		for _, sub := range sc.cache {
+			if oldestSub == nil {
+				oldestSub = sub
+				continue
+			}
+
+			if sub.lastAccessed < oldestSub.lastAccessed {
+				oldestSub = sub
+			}
+		}
+
+		if oldestSub != nil {
+			delete(sc.cache, oldestSub.na.IP.String())
+		}
+	}
+
+	sub.score = 1
+	sub.lastAccessed = time.Now().Unix()
+	sc.cache[key] = sub
+	return nil
+}
+
+// exists returns true if the provided key exist in the submissions cache.
+func (sc *naSubmissionCache) exists(key string) bool {
+	sc.mtx.Lock()
+	defer sc.mtx.Unlock()
+
+	if key == "" {
+		return false
+	}
+
+	_, ok := sc.cache[key]
```

Comment:
`defer` has fairly high overhead.  For simple cases, it's more efficient to just lock and unlock around the critical section.
```Go
if key == "" {
	return false
}

sc.mtx.Lock()
_, ok := sc.cache[key]
sc.mtx.Unlock()
return ok
```